src/audio/dai-zephyr.c

```diff
@@ -0,0 +1,1408 @@
+        if (!group->num_dais)
+                group->group_id = 0;
+}
+
+/* Assign DAI to a group */
+int dai_assign_group(struct comp_dev *dev, uint32_t group_id)
+{
+	struct dai_data *dd = comp_get_drvdata(dev);
+
+	if (dd->group) {
+		if (dd->group->group_id != group_id) {
+			comp_err(dev, "dai_assign_group(), DAI already in group %d, requested %d",
+				 dd->group->group_id, group_id);
+			return -EINVAL;
+		}
+
+		/* No need to re-assign to the same group, do nothing */
+		return 0;
+	}
+
+	dd->group = dai_group_get(group_id, DAI_CREAT);
+	if (!dd->group) {
+		comp_err(dev, "dai_assign_group(), failed to assign group %d",
+			 group_id);
+		return -EINVAL;
+	}
+
+	comp_dbg(dev, "dai_assign_group(), group %d num %d",
+		 group_id, dd->group->num_dais);
+
+	/* Register for the atomic trigger event */
+	notifier_register(dev, dd->group, NOTIFIER_ID_DAI_TRIGGER,
+			  dai_atomic_trigger, 0);
+
+	return 0;
+}
+
+struct dai *dai_get(uint32_t type, uint32_t index, uint32_t flags)
+{
+	const struct dai_type_info *dti;
+	struct dai *d;
+
+	tr_info(&dai_comp_tr, "log_ssp_data get_dai_zephyr index: %u", index);
+
+	dti = dai_find_type(type);
+	if (!dti)
+		return NULL; /* type not found */
+
+	for (d = dti->dai_array; d < dti->dai_array + dti->num_dais; d++) {
+		if (d->index == index)
+			return dai_probe(d->drv) ? NULL : d;
+	}
+
+	return NULL;
+}
+
+void dai_put(struct dai *dai)
+{
+	int ret;
+
+	ret = dai_remove(dai->drv);
+	if (ret < 0) {
+		tr_err(&dai_comp_tr, "dai_put_zephyr: index %d failed ret = %d",
+		       dai->index, ret);
+	}
+}
+
+
+int dai_init(struct sof *sof)
+{
+	struct dai *dai;
+	int i;
+
+#if CONFIG_INTEL_SSP
+	const struct dai_type_info *ssp = NULL;
+	char ssp_name[] = "SSP_X";
+#endif
+#if CONFIG_INTEL_ALH
+	const struct dai_type_info *alh = NULL;
+#endif
+
+	for (i = 0; i < ARRAY_SIZE(dti); i++)
+		switch (dti[i].type) {
+#if CONFIG_INTEL_SSP
+		case SOF_DAI_INTEL_SSP:
+			ssp = dti + i;
+			break;
+#endif
+#if CONFIG_INTEL_ALH
+		case SOF_DAI_INTEL_ALH:
+			alh = dti + i;
+			break;
+#endif
+		}
+
+#if CONFIG_INTEL_SSP
+	dai = ssp->dai_array;
+
+	/* init ssp */
+	for (i = 0; i < ssp->num_dais; i++) {
+		/* name varied by zephyr device tree index */
+		dai->index = i;
+		dai->type = SOF_DAI_INTEL_SSP;
+		dai->dma_caps = DMA_CAP_GP_LP | DMA_CAP_GP_HP;
+		dai->dma_dev = DMA_DEV_SSP;
+		ssp_name[4] = (uint8_t)((i % 10UL) + '0');
+		dai->drv = device_get_binding(ssp_name);
+		if (!dai->drv)
+			return -EINVAL;
+		dai++;
+	}
+#endif
+
+#if CONFIG_INTEL_ALH
+	dai = alh->dai_array;
+
+	for (i = 0; i < alh->num_dais; i++) {
+		dai->index = (i / DAI_NUM_ALH_BI_DIR_LINKS_GROUP) << 8 |
+			(i % DAI_NUM_ALH_BI_DIR_LINKS_GROUP);
+		dai->type = SOF_DAI_INTEL_ALH;
+		dai->dma_caps = DMA_CAP_GP_LP | DMA_CAP_GP_HP;
+		dai->dma_dev = DMA_DEV_SSP;
+		switch (i) {
+		case 0:
+			dai->drv = device_get_binding("ALH_0");
+			break;
+		case 1:
+			dai->drv = device_get_binding("ALH_1");
```

Comment:
would it be possible to do the same as with DDP names to (1) use the same code and (2) be flexible even if more than 2 ALHs can never happen?